### ISSUE TITLE
[IMPROVEMENT] Speed Up MiniMax Algorithm

### ISSUE THREAD

**Issue by mzimm003** [mzimm003/Chess#7]
# Is your feature request related to a problem? Please describe.
As is, the MiniMax algorithm (class my_chess.learner.models.DeepChessAlphaBeta) is barely able to recurse on 3 moves in a timely fashion. It is unclear, but possible, this is the cause for poor performance even with a board evaluator providing 99% accuracy.

# Describe the solution you'd like

1. Testing to determine the timings of the elements of the algorithm should be implemented. This will help target what causes the greatest slowdown.
2.  Depending on the result of testing, improvements could include:

     1.  minimax enhancements (e.g. hash table to avoid recalculating previously seen positions), 
     2. exposing and fixing poorly implemented existing infrastructure
     3. providing smaller, distilled models for board evaluation

# Describe alternatives you've considered
Distilled models have been attempted, but with seemingly no impact on timings, making the exact bottleneck unclear.

It is also possible the dataset split is creating a false sense of model efficacy. The split is done randomly by observation, rather than by game. The training set and validation set may have too much in common then, causing an overly optimistic 99% accuracy score when measuring which board is winning and which is losing. If this is true, it is possible it is unnecessary for the minimax algorithm to go any deeper to be effective.

# Additional context
The Minimax algorithm is based on the DeepChess research paper, which makes use of alpha beta pruning, though stores a board state instead of a valuation. This has been implemented already, but the hash table also mentioned in the paper has not. It is unclear from the paper what depths were achieved in their search, but elo ratings achieved by their bot suggest the search did significantly better than this repo's current implementation.
